cognite/client/data_classes/assets.py

```diff
@@ -449,3 +456,244 @@ def dump(self, camel_case: bool = False) -> Dict[str, Any]:
+
+    def groupby_parent_xid(self) -> Dict[Optional[str], List[Asset]]:
+        self.is_valid(on_error="raise")
+        # Sort (on parent) as required by groupby. Since the empty string "" is a valid external ID,
+        # we sort based on the hash of the parent - which may be None:
+        parent_sorted = sorted(self._assets, key=lambda a: hash(self._get_parent(a)))
+        return {
+            parent: list(child_assets)
+            for parent, child_assets in itertools.groupby(parent_sorted, key=self._get_parent)
+        }
+
+    def _inspect_attributes(self) -> Tuple[List[Asset], List[Asset], List[Asset], Dict[str, List[Asset]]]:
+        missing, orphans, roots, duplicates = [], [], [], defaultdict(list)
+        xid_count = Counter(map(self._get_xid, self._assets))
+
+        for asset in self._assets:
+            xid, parent, name = asset.external_id, asset.parent_external_id, asset.name
+            if xid is None or name is None or len(name) < 1:
+                missing.append(asset)
+                continue  # Don't report missing as part of any other group
+
+            if xid_count[xid] > 1:
+                duplicates[xid].append(asset)
+
+            if parent is None:
+                roots.append(asset)
+            elif parent not in xid_count:
+                orphans.append(asset)
+
+        return roots, orphans, missing, dict(duplicates)
+
+    def _preconditions_for_cycle_check_are_met(self, on_error: OnErrorOptions) -> bool:
+        # Required before cycle-check can run
+        if not (not self._roots or self._orphans or self._missing or self._duplicates):
+            return True
+        self._on_error(on_error, "Unable to run cycle-check before basic issues are fixed.")
+        return False
+
+    def _locate_cycles(self, on_error: OnErrorOptions) -> Tuple[int, List[List[str]]]:
+        edges = dict(map(op.attrgetter("external_id", "parent_external_id"), self._assets))
+        has_cycles, no_cycles = set(), {None, *map(self._get_xid, self.roots)}
+
+        for xid, parent in edges.items():
+            if parent in no_cycles:
+                no_cycles.add(xid)
+
+            elif parent in has_cycles:
+                has_cycles.add(xid)
+
+            elif xid == parent:
+                has_cycles.add(xid)
+
+            else:
+                counter, seen = itertools.count(), {xid}
+                while next(counter) < 1_000_000:
+                    xid, parent = parent, edges.get(parent)
+                    seen.add(xid)
+                    if parent in no_cycles:
+                        no_cycles |= seen
+                        break
+                    elif parent in seen:
+                        has_cycles |= seen
+                        break
+                else:  # escape hatch if we don't break out of while
+                    raise RuntimeError("Failed to locate root asset or a cycle after 1 mill. repeated parent lookups")
```

Comment:
RuntimeError(f"Failed to locate root asset or a cycle after {max_hierarchy_depth} repeated parent lookups")